File: src/main/java/org/wysko/midis2jam2/instrument/brass/Tuba.java
package org.wysko.midis2jam2.instrument.brass;

import com.jme3.material.Material;
import com.jme3.math.Quaternion;
import com.jme3.scene.Node;
import org.jetbrains.annotations.NotNull;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.AnimatedKeyCloneByIntegers;
import org.wysko.midis2jam2.instrument.Axis;
import org.wysko.midis2jam2.instrument.MonophonicInstrument;
import org.wysko.midis2jam2.instrument.PressedKeysFingeringManager;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;

import java.util.Arrays;
import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;

/**
 * The tuba.
 */
public class Tuba extends MonophonicInstrument {
	
	public static final PressedKeysFingeringManager FINGERING_MANAGER = PressedKeysFingeringManager.from(Tuba.class);
	
	public Tuba(Midis2jam2 context, List<MidiChannelSpecificEvent> eventList) throws ReflectiveOperationException {
		super(
				context,
				eventList,
				TubaClone.class,
				FINGERING_MANAGER
		);
		
		groupOfPolyphony.setLocalTranslation(-71, 28.9f, -97.9f);
	}
	
	@Override
	protected void moveForMultiChannel() {
		offsetNode.setLocalTranslation(0, 40 * indexForMoving(), 0);
	}
	
	public class TubaClone extends AnimatedKeyCloneByIntegers {
		
		public TubaClone() {
			super(Tuba.this, -0.05f, 0.8f, 4, Axis.Y, Axis.Z);
			
			body = context.loadModel("TubaBody.fbx", "HornSkin.bmp", Midis2jam2.MatType.REFLECTIVE, 0.9f);
			bell.attachChild(context.loadModel("TubaHorn.obj", "HornSkin.bmp", Midis2jam2.MatType.REFLECTIVE, 0.9f));
			modelNode.attachChild(body);
			modelNode.attachChild(bell);
			Material material = context.reflectiveMaterial("Assets/HornSkinGrey.bmp");
			((Node) body).getChild(1).setMaterial(material);
			
			for (int i = 0; i < 4; i++) {
				keys[i] = context.loadModel("TubaKey" + (i + 1) + ".obj", "HornSkinGrey.bmp",
						Midis2jam2.MatType.REFLECTIVE, 0.9f);
				modelNode.attachChild(keys[i]);
			}
			
			idleNode.setLocalRotation(new Quaternion().fromAngles(rad(-10), rad(90), 0));
			highestLevel.setLocalTranslation(10, 0, 0);
		}
		
		@Override
		protected void moveForPolyphony() {
			offsetNode.setLocalRotation(new Quaternion().fromAngles(0, rad(50 * indexForMoving()), 0));
		}
		
		@Override
		protected void animateKeys(@NotNull Integer[] pressed) {
			/* Trumpet keys move down when pressed */
			for (int i = 0; i < 4; i++) {
				int finalI = i;
				if (Arrays.stream(pressed).anyMatch(integer -> integer == finalI)) {
					keys[i].setLocalTranslation(0, -0.5f, 0);
				} else {
					keys[i].setLocalTranslation(0, 0, 0);
				}
			}
		}
	}
}

File: app/src/commonMain/kotlin/org/wysko/midis2jam2/instrument/family/brass/Tuba.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.brass

import com.jme3.math.Vector3f
import com.jme3.scene.Node
import com.jme3.scene.Spatial
import org.wysko.kmidi.midi.event.MidiEvent
import org.wysko.midis2jam2.manager.PerformanceManager
import org.wysko.midis2jam2.instrument.MonophonicInstrument
import org.wysko.midis2jam2.instrument.MultipleInstancesLinearAdjustment
import org.wysko.midis2jam2.instrument.algorithmic.PressedKeysFingeringManager
import org.wysko.midis2jam2.instrument.clone.ClonePitchBendConfiguration
import org.wysko.midis2jam2.instrument.clone.CloneWithKeyPositions
import org.wysko.midis2jam2.util.*
import org.wysko.midis2jam2.world.Axis
import org.wysko.midis2jam2.world.assetLoader
import org.wysko.midis2jam2.world.modelR
import kotlin.time.Duration

private val FINGERING_MANAGER: PressedKeysFingeringManager = PressedKeysFingeringManager.from(Tuba::class)

/**
 * The Tuba.
 */
class Tuba(context: PerformanceManager, eventList: List<MidiEvent>) :
    MonophonicInstrument(context, eventList, TubaClone::class, FINGERING_MANAGER), MultipleInstancesLinearAdjustment {

    override val pitchBendConfiguration: ClonePitchBendConfiguration = ClonePitchBendConfiguration(Axis.Z)
    override val multipleInstancesDirection: Vector3f = v3(0, 40, 0)

    init {
        placement.loc = v3(-110f, 25f, -30f)
    }

    /**
     * A single Tuba.
     */
    inner class TubaClone : CloneWithKeyPositions(this@Tuba, -0.05f, 0.8f, Axis.Y, Axis.X) {

        override val keys: Array<Spatial> = Array(4) { i ->
            with(geometry) {
                +context.modelR("TubaKey${i + 1}.obj", "HornSkinGrey.bmp")
            }
        }

        override fun adjustForPolyphony(delta: Duration) {
            root.rot = v3(0, 50, 0) * indexForMoving()
        }

        override fun animateKeys(pressed: List<Int>) {
            super.animateKeys(pressed)
            keys.forEachIndexed { i, key -> key.loc = v3(0f, if (i + 1 in pressed) -0.5 else 0, 0f) }
        }

        init {
            with(geometry) {
                +context.modelR("TubaBody.obj", "HornSkin.bmp").apply {
                    (this as Node)[1].material = context.assetLoader.reflectiveMaterial("Assets/HornSkinGrey.bmp")
                }
            }
            with(bell) {
                +context.modelR("TubaHorn.obj", "HornSkin.bmp")
            }

            highestLevel.loc = v3(10, 0, 0)
            highestLevel.rot = v3(-10.0, 90.0, 0)
        }
    }
}
